gbn/timeout_manager.go

```diff
@@ -0,0 +1,522 @@
+			// update the resend timeout when we receive the
+			// corresponding response.
+			delete(m.sentTimes, msg.Seq)
+
+			m.resendBooster.Boost()
+
+			return
+		}
+
+		m.sentTimes[msg.Seq] = sentAt
+	}
+}
+
+// Received should be called when a message is received by the connection.
+func (m *TimeoutManager) Received(msg Message) {
+	if m.useStaticTimeout {
+		return
+	}
+
+	receivedAt := time.Now()
+
+	switch msg := msg.(type) {
+	case *PacketSYN, *PacketSYNACK:
+		m.latestSentSYNTimeMu.Lock()
+
+		if m.latestSentSYNTime.IsZero() {
+			m.latestSentSYNTimeMu.Unlock()
+
+			return
+		}
+
+		responseTime := receivedAt.Sub(m.latestSentSYNTime)
+
+		m.latestSentSYNTime = time.Time{}
+
+		m.mu.Lock()
+		defer m.mu.Unlock()
+
+		m.latestSentSYNTimeMu.Unlock()
+
+		m.updateResendTimeoutUnsafe(responseTime)
+
+	case *PacketACK:
+		m.sentTimesMu.Lock()
+
+		sentTime, ok := m.sentTimes[msg.Seq]
+		if !ok {
+			m.sentTimesMu.Unlock()
+
+			return
+		}
+
+		delete(m.sentTimes, msg.Seq)
+
+		m.mu.Lock()
+		defer m.mu.Unlock()
+
+		m.sentTimesMu.Unlock()
+
+		m.responseCounter++
+
+		reachedFrequency := m.responseCounter%
+			m.timeoutUpdateFrequency == 0
+
+		// In case we never set the resend timeout dynamically in the
+		// handshake due to needing to resend the SYN, or if we've
+		// reached received the number of packages matching the
+		// timeoutUpdateFrequency, we'll update the resend timeout.
+		if !m.hasSetDynamicTimeout || reachedFrequency {
+			m.responseCounter = 0
+
+			m.updateResendTimeoutUnsafe(receivedAt.Sub(sentTime))
+		}
+	}
+}
+
+// updateResendTimeout updates the resend timeout based on the given response
+// time. The resend timeout will be only be updated if the given response time
+// is greater than the default resend timeout, after being multiplied by the
+// resendMultiplier.
+//
+// NOTE: This function TimeoutManager mu must be held when calling this
+// function.
+func (m *TimeoutManager) updateResendTimeoutUnsafe(responseTime time.Duration) {
+	m.hasSetDynamicTimeout = true
+
+	multipliedTimeout := time.Duration(m.resendMultiplier) * responseTime
+
+	if multipliedTimeout < minimumResendTimeout {
+		m.log.Tracef("Setting resendTimeout to minimumResendTimeout "+
+			"%v as the new dynamic timeout %v is not greater than "+
+			"the minimum resendTimeout.",
+			m.resendTimeout, multipliedTimeout)
+		multipliedTimeout = minimumResendTimeout
+	}
+
+	m.log.Tracef("Updating resendTimeout to %v", multipliedTimeout)
+
+	m.resendTimeout = multipliedTimeout
+
+	// Also update and reset the resend booster, as the new dynamic
+	// resend timeout most accurately reflects the current response
+	// time.
+	m.resendBooster.Reset(multipliedTimeout)
+
+	// As we may have received a data packet that executes this function
+	// while we are also concurrently resending the queue, we also restart
+	// the frequency timeout, to ensure that the messages we're resending
+	// won't boost the resend timeout.
+	m.resendBooster.RestartFrequencyTimeout()
```

Comment:
one last thing:

these two are both only ever called together & are only ever called here. But each of them acquire the booster's mutex individually. So why not instead compile the two? `ResetAndRestart` or something like that? then that lock only needs to be acquired once

Reply:
Sure! I opted to just include the `RestartFrequencyTimeout` in the `Reset` function, but instead keep `Reset` name with an added comment. Hope you agree about that approach :)!